### ISSUE TITLE
MSVC warning 4146 on radix sort

### ISSUE THREAD

**Issue by gamemachine** [madmann91/bvh#34]
This was on VS 16.9.5 not sure if it's harmless or not, I silenced it with a #pragma.

**Comment by madmann91** [comment]
Which warning? And where? Without more information I cannot solve this.

**Comment by gamemachine** [comment]
Sorry that's C4146 on the error.
in radix_sort.hpp

**Comment by madmann91** [comment]
Which line?

**Comment by madmann91** [comment]
Also I don't have MSVC

**Comment by gamemachine** [comment]
C4146 is the actual content of the error.   'unary minus operator applied to unsigned type, result still unsigned'.

It's line 93 looks like the -y triggers it.

**Comment by madmann91** [comment]
This is perfectly fine. The C++ standard states:
```
A computation  involving  unsigned  operands  can  never overflow, because  a  result  that  cannot
be  represented  by  the  resulting  unsigned  integer  type  is reduced  modulo  the  number  that  is
one  greater  than  the  largest  value  that  can  be represented by the resulting type.
```
So it's MSVC being stupid in its warnings. Negating unsigned numbers is of course not so common, but it's perfectly defined and has a precise meaning. Here it's used in order to make turn a floating point value into a key that can be sorted using integer radix sort. The only assumption there is that floats are represented with IEEE-754, which is reasonable I think.

**Comment by kostinio** [comment]
@madmann91, don't you want to add #pragma warning ( disable : 4146 ) at the top of the file? This warning is threated as error and compilation fails.

**Comment by madmann91** [comment]
I don't want to enable/disable warnings like this. MSVC has a lot of spurious warnings anyway and who knows which ones will trigger the next time I touch this code (and also which ones I'll have to disable once I update the code if they no longer apply). However, I also see that this can be annoying if you compile with warnings as errors (which I wouldn't recommend in general -- but you do you), so I can think of the following compromise: Try the patch below, if it works create a PR and I'll merge it. If not, well, try something in the same vein, basically negating a signed integer and putting that back into an unsigned one.

```cpp
// replace this version...
static typename SizedIntegerType<sizeof(T) * CHAR_BIT>::Unsigned make_key(T x) {
    using U = typename SizedIntegerType<sizeof(T) * CHAR_BIT>::Unsigned;
    auto mask = U(1) << (sizeof(T) * CHAR_BIT - 1);
    auto y = as<U>(x);
    return (y & mask ? (-y) ^ mask : y) ^ mask;
}
// ...with that version
static typename SizedIntegerType<sizeof(T) * CHAR_BIT>::Unsigned make_key(T x) {
    using U = typename SizedIntegerType<sizeof(T) * CHAR_BIT>::Unsigned;
    using I = typename SizedIntegerType<sizeof(T) * CHAR_BIT>::Signed;
    auto mask = U(1) << (sizeof(T) * CHAR_BIT - 1);
    auto y = as<U>(x);
    return (y & mask ? static_cast<U>(-static_cast<I>(y)) ^ mask : y) ^ mask;
}
```

**Comment by kostinio** [comment]
Looks like now it works same way but without warning. I checked it with numbers which are close to 2^n (sign bit is used). And also compared traverse statistic - it was the same.

**Comment by kostinio** [comment]
How do you make PR here? I can't create branch, but i see merged PRs like that:
```
madmann91 merged 1 commit into madmann91:master from newr5:master
```
I have no idea how to do this trick.

**Comment by madmann91** [comment]
You can click the fork button (top right when on the project page), then this will create your own fork of the project. Then you can commit into it via normal means. Once that is done, you can go on to your fork in the web browser and create a pull request from there. I'll review it and merge it. Once that is done then you can delete your fork. Thank you for putting that extra effort!

**Comment by kostinio** [comment]
Done! Thank you for cooperation.

**Comment by madmann91** [comment]
Should be fixed by commit 93a7cdc
